Add manual pages (#26)
NLnetLabs/dnst: doc/manual/source/man/dnst-signzone.rst

dnst-signzone
===============

Synopsis
--------

:program:`dnst signzone` [``OPTIONS``] ``<ZONEFILE>`` ``<KEY>...``

Description
-----------

> **dnst signzone** signs the zone with the given key(s).

Reviewer: For consistency and accuracy I think this should say `zonefile` like we use below.

Reviewer 2: I guess technically it signs the _zone in the zonefile_?

Keys must be specified by their base name (usually ``K<name>+<alg>+<id>``),
i.e. WITHOUT the ``.private`` or ``.key`` extension. Both ``.private`` and
``.key`` files are required.

A date can be a timestamp (seconds since the epoch), or of the form

> <YYYYMMdd[hhmmss]>.

Reviewer 2: ```suggestion
``<YYYYMMdd[hhmmss]>``.
```

Options
-------

.. option:: <ZONEFILE>

      The zonefile to sign.

.. option:: <KEY>...

      The keys to sign the zone with.

.. option:: -b

>       Use a more readable layout in the signed zone file and print comments on

Reviewer: The layout of `dnst` output is NOT affected by this option, it only adds comments. The original `ldns-signzone` sets an internal flag `LDNS_COMMENT_LAYOUT` which has an effect in various places in the ldns library code on adding yet more comments (e.g. `; <void>`, `;; Empty non-terminal: <name>` and `;; Zone: <zone name>`) and adds some line breaks, but these effects have not been completely replicated in the `dnst` version yet.

Reviewer: Further, per the original `ldns-signzone` man page: "Without this option, only DNSKEY RR's will have their Key Tag annotated in the comment text." - this IS implemented, but reads from the current `dnst` man page text as if no comments are ouput unless `-b` is used which is not true.

      DNSSEC records.

.. option:: -d

>       Do not add used keys to the resulting zone file.

Reviewer: ```suggestion
      Do not add used keys to the resulting zonefile.
```

.. option:: -e <DATE>

>       Set the expiration date. Defaults to 4 weeks from now.

Reviewer: As is said in the original man page: "of signatures".

.. option:: -f <FILE>

      Write signed zone to file. Use ``-f -`` to output to stdout. Defaults to
      ``<ZONEFILE>.signed``.

.. option:: -i <DATE>

>       Set the inception date. Defaults to now.

Reviewer: As is said in the original man page: "of signatures".

> .. option:: -o <domain>

Reviewer: ```suggestion
.. option:: -o <DOMAIN>
```

>       Set the origin for the zone (for zonefiles with relative names and no

Reviewer: This is correct but perhaps it would be better to say that this is only needed for zones that both contain relative labels AND lack an `$ORIGIN` directive.

Author: Yes. It needs to be an absolute domain name. Also, signzone or rather domain's Name (::from_str?) turns any non-dot-terminated domain into an absolute/dot-terminated domain. So `dnst signzone -o origin` and `dnst signzone -o origin.` do the same.

      $ORIGIN).

.. option:: -u

      Set SOA serial to the number of seconds since Jan 1st 1970.

      If this would NOT result in the SOA serial increasing it will be
      incremented instead.

.. option:: -n

>       Use NSEC3 instead of NSEC. If specified, you can use extra options (see
>       :ref:`dnst-signzone-nsec3-options`).

Reviewer: ```suggestion
      Use NSEC3 instead of NSEC. By default RFC 9276 best practice settings are used: SHA-1, no extra iterations, empty salt. To use different NSEC3 settings see :ref:`dnst-signzone-nsec3-options`.
```

.. option:: -H

      Hash only, don't sign.

.. option:: -h, --help

      Print the help text (short summary with ``-h``, long help with
      ``--help``).

.. _dnst-signzone-nsec3-options:

NSEC3 options
--------------------------------

NSEC3 options for use with ``-n``.

> .. option:: -a <ALGORITHM>

Reviewer: ```suggestion
.. option:: -a <ALGORITHM NUMBER OR MNEMONIC>
```

Reviewer: I wonder if we shouldn't remove this argument entirely from `dnst` as the only supported algorithm is SHA-1.

Author: Yeah, might be a good idea, as it serves no real purpose currently.

      Specify the hashing algorithm. Defaults to SHA-1.

.. option:: -t <NUMBER>

>       Set the number of hash iterations. Defaults to 0.

Reviewer: ```suggestion
      Set the number of extra hash iterations. Defaults to 0.
```

Reviewer 2: This feels so wrong (the behaviour, not the description)

.. option:: -s <STRING>

>       Specify the salt. Defaults to ``-``, meaning no salt.

Reviewer: I think there needs to be some mention here that the string must be hex encoded.

Reviewer: ```suggestion
      Specify the salt. Defaults to ``-``, meaning empty salt.
```

.. option:: -i <DATE>

      Set the inception date of signatures to this date. Defaults to now.

.. option:: -o <DOMAIN>

      Set the origin for the zone (only necessary for zonefiles with relative
      names and no $ORIGIN).

.. option:: -u

      Set SOA serial to the number of seconds since Jan 1st 1970.

      If this would NOT result in the SOA serial increasing it will be
      incremented instead.

.. option:: -n

      Use NSEC3 instead of NSEC. By default, RFC 9276 best practice settings
      are used: SHA-1, no extra iterations, empty salt. To use different NSEC3
      settings see :ref:`dnst-signzone-nsec3-options`.

.. option:: -H

      Hash only, don't sign.

.. option:: -h, --help

      Print the help text (short summary with ``-h``, long help with
      ``--help``).

.. _dnst-signzone-nsec3-options:

NSEC3 options
--------------------------------

The following options can be used with ``-n`` to override the default NSEC3
settings used.

.. option:: -a <ALGORITHM NUMBER OR MNEMONIC>

      Specify the hashing algorithm. Defaults to SHA-1.

.. option:: -t <NUMBER>

      Set the number of hash iterations. Defaults to 0.

.. option:: -s <STRING>

>       Specify the salt as a hex string. Defaults to ``-``, meaning no salt.

Reviewer: ```suggestion
      Specify the salt as a hex string. Defaults to ``-``, meaning empty salt.
```
